p2p/transport/webrtc/connection.go

```diff
@@ -0,0 +1,242 @@
+	localPeer      peer.ID
+	privKey        ic.PrivKey
+	localMultiaddr ma.Multiaddr
+
+	remotePeer      peer.ID
+	remoteKey       ic.PubKey
+	remoteMultiaddr ma.Multiaddr
+
+	m       sync.Mutex
+	streams map[uint16]*dataChannel
+
+	streamChan chan network.MuxedStream
+
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+func newConnection(
+	pc *webrtc.PeerConnection,
+	transport *WebRTCTransport,
+	scope network.ConnManagementScope,
+
+	localPeer peer.ID,
+	privKey ic.PrivKey,
+	localMultiaddr ma.Multiaddr,
+
+	remotePeer peer.ID,
+	remoteKey ic.PubKey,
+	remoteMultiaddr ma.Multiaddr,
+) *connection {
+	streamChan := make(chan network.MuxedStream, maxStreamBufferSize)
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	conn := &connection{
+		pc:        pc,
+		transport: transport,
+		scope:     scope,
+
+		localPeer:      localPeer,
+		privKey:        privKey,
+		localMultiaddr: localMultiaddr,
+
+		remotePeer:      remotePeer,
+		remoteKey:       remoteKey,
+		remoteMultiaddr: remoteMultiaddr,
+		ctx:             ctx,
+		cancel:          cancel,
+		streams:         make(map[uint16]*dataChannel),
+
+		streamChan: streamChan,
+	}
+
+	pc.OnDataChannel(func(dc *webrtc.DataChannel) {
+		id := *dc.ID()
+		var stream *dataChannel
+		dc.OnOpen(func() {
+			// datachannel cannot be detached before opening
+			rwc, err := dc.Detach()
+			if err != nil {
+				log.Errorf("[%s] could not detch channel: %s", localPeer, dc.Label())
+				return
+			}
+			stream = newDataChannel(conn, dc, rwc, pc, nil, nil)
+			conn.addStream(id, stream)
+			streamChan <- stream
+		})
+
+	})
+
+	pc.OnConnectionStateChange(func(state webrtc.PeerConnectionState) {
+		if state == webrtc.PeerConnectionStateClosed || state == webrtc.PeerConnectionStateDisconnected {
+			conn.Close()
+		}
+	})
+
+	return conn
+}
+
+// ConnState implements transport.CapableConn
+func (c *connection) ConnState() network.ConnectionState {
+	return network.ConnectionState{}
+}
+
+// Implement network.MuxedConn
+
+func (c *connection) Close() error {
+	if c.IsClosed() {
+		return nil
+	}
+
+	c.scope.Done()
+	c.cancel()
+	_ = c.pc.Close()
+	return nil
+}
+
+func (c *connection) IsClosed() bool {
+	select {
+	case <-c.ctx.Done():
+		return true
+	default:
+	}
+	return false
+}
+
+func (c *connection) OpenStream(ctx context.Context) (network.MuxedStream, error) {
+	type openStreamResult struct {
+		network.MuxedStream
+		error
+	}
+
+	if c.IsClosed() {
+		return nil, os.ErrClosed
+	}
+
+	result := make(chan openStreamResult)
+	dc, err := c.pc.CreateDataChannel("", nil)
+	if err != nil {
+		return nil, err
+	}
+
+	streamID := *dc.ID()
+	var stream *dataChannel
+	dc.OnOpen(func() {
+		rwc, err := dc.Detach()
+		if err != nil {
+			result <- openStreamResult{
+				nil,
+				errDatachannel("could not detach", err),
+			}
+			return
+		}
+		stream = newDataChannel(c, dc, rwc, c.pc, nil, nil)
+		c.addStream(streamID, stream)
+		result <- openStreamResult{stream, err}
+	})
+
+	dc.OnClose(func() {
```

Comment:
I've gotten rid of the `OnClose` callbacks since we can detect when a connection is closed only when `Read` is called on the datachannel.

Reply:
Not sure I understand your comment here, because:
1. It's still here.
1. This is the `OnClose` callback of the datachannel, not of the connection.

Reply:
It shouldn't be there in the latest commit. The datachannel needs to read on the underlying sctp stream to know if the remote has closed. Since we used a detached datachannel (we have fewer abstractions over the sctp stream), we will only receive that message once we issue a `Read` call to the datachannel. We only issue a read call after `OnOpen` is called, and referring to [pion webrtc datachannel](https://github.com/pion/webrtc/blob/7ab3174640b3ce15abebc2516a2ca3939b5f105f/datachannel.go#L282), the `OnOpen` callback fires immediately for detached channels.